src/server/journal/types.h

```diff
@@ -35,30 +37,25 @@ struct Entry : public EntryBase {
                    std::pair<std::string_view, ArgSlice>  // Command and its shard parts.
                    >;
 
-  Entry(TxId txid, DbIndex dbid, Payload pl, uint32_t shard_cnt)
-      : EntryBase{txid, journal::Op::COMMAND, dbid, shard_cnt}, payload{pl} {
+  Entry(TxId txid, Op opcode, DbIndex dbid, uint32_t shard_cnt, Payload pl)
+      : EntryBase{txid, opcode, dbid, shard_cnt}, payload{pl} {
   }
 
   Entry(journal::Op opcode, DbIndex dbid) : EntryBase{0, opcode, dbid, 0}, payload{} {
   }
+  Entry(TxId txid, journal::Op opcode, DbIndex dbid, uint32_t shard_cnt)
+      : EntryBase{txid, opcode, dbid, shard_cnt}, payload{} {
+  }
 
   Payload payload;
 };
 
 struct ParsedEntry : public EntryBase {
-  // Payload represents the parsed command.
-  using Payload = std::optional<CmdArgVec>;
-
-  ParsedEntry() = default;
-
-  ParsedEntry(journal::Op opcode, DbIndex dbid) : EntryBase{0, opcode, dbid, 0}, payload{} {
-  }
-
-  ParsedEntry(TxId txid, DbIndex dbid, Payload pl, uint32_t shard_cnt)
-      : EntryBase{txid, journal::Op::COMMAND, dbid, shard_cnt}, payload{pl} {
-  }
-
-  Payload payload;
+  struct CmdData {
+    std::unique_ptr<std::string> command_buf;
+    CmdArgVec cmd_args;  // represents the parsed command.
```

Comment:
Double indirection... Is this a guard against accidentally copying?

Then we can make the whole struct non movable or just use a std::unique<char[]> (and see my suggested reader optmizations, so we just need to copy the array)